notify/sns/sns.go

```diff
@@ -0,0 +1,171 @@
+)
+
+// Notifier implements a Notifier for SNS notifications.
+type Notifier struct {
+	conf    *config.SNSConfig
+	tmpl    *template.Template
+	logger  log.Logger
+	client  *http.Client
+	retrier *notify.Retrier
+}
+
+func (n Notifier) Notify(ctx context.Context, alert ...*types.Alert) (bool, error) {
+	credentials := credentials.NewStaticCredentials(n.conf.Sigv4.AccessKey, string(n.conf.Sigv4.SecretKey), "")
+	if n.conf.Sigv4.AccessKey == "" {
+		credentials = nil
+	}
+
+	sess, err := session.NewSessionWithOptions(session.Options{
+		Config: aws.Config{
+			Region:      aws.String(n.conf.Sigv4.Region),
+			Credentials: credentials,
+			Endpoint:    aws.String(n.conf.APIUrl),
+		},
+		Profile: n.conf.Sigv4.Profile,
+	})
+
+	if _, err := sess.Config.Credentials.Get(); err != nil {
+		return false, fmt.Errorf("could not get SigV4 credentials: %w", err)
+	}
+
+	if n.conf.Sigv4.RoleARN != "" {
+		sess.Config.Credentials = stscreds.NewCredentials(sess, n.conf.Sigv4.RoleARN)
+	}
+
+	data := notify.GetTemplateData(ctx, n.tmpl, alert, n.logger)
+	tmpl := notify.TmplText(n.tmpl, data, &err)
+	message := tmpl(n.conf.Message)
+
+	client := sns.New(sess, &aws.Config{Credentials: credentials})
+	publishInput := &sns.PublishInput{}
+
+	if n.conf.TopicARN != "" {
+		publishInput.SetTopicArn(n.conf.TopicARN)
+		messageToSend, isTrunc, err := validateAndTruncateMessage(message)
+		if err != nil {
+			return false, err
+		}
+		if isTrunc {
+			n.conf.Attributes["truncated"] = "true"
+		}
+		publishInput.SetMessage(messageToSend)
+	}
+	if n.conf.PhoneNumber != "" {
+		publishInput.SetPhoneNumber(n.conf.PhoneNumber)
+		// If SMS message is over 1600 chars, SNS will reject the message.
+		_, isTruncated := notify.Truncate(message, 1600)
+		if isTruncated {
+			return false, fmt.Errorf("SMS message exeeds length of 1600 charactors")
+		} else {
+			publishInput.SetMessage(message)
+		}
+	}
+	if n.conf.TargetARN != "" {
+		publishInput.SetTargetArn(n.conf.TargetARN)
+		messageToSend, isTrunc, err := validateAndTruncateMessage(message)
+		if err != nil {
+			return false, err
+		}
+		if isTrunc {
+			n.conf.Attributes["truncated"] = "true"
+		}
+		publishInput.SetMessage(messageToSend)
+	}
+
+	if len(n.conf.Attributes) > 0 {
+		attributes := map[string]*sns.MessageAttributeValue{}
+		for k, v := range n.conf.Attributes {
+			attributes[k] = &sns.MessageAttributeValue{DataType: aws.String("String"), StringValue: aws.String(v)}
+		}
+		publishInput.SetMessageAttributes(attributes)
+	}
+
+	if n.conf.Subject != "" {
+		publishInput.SetSubject(n.conf.Subject)
+	}
+
+	// Deduplication key is only added if it's a FIFO SNS Topic.
+	if n.conf.IsFIFOTopic {
+		key, err := notify.ExtractGroupKey(ctx)
+		if err != nil {
+			return false, err
+		}
+		publishInput.SetMessageDeduplicationId(key.Hash())
+	}
+
+	publishOutput, err := client.Publish(publishInput)
+	if err != nil {
+		// AWS Response is bad, probably a config issue.
+		return false, err
```

Comment:
Why `return false` (no retry)? I think the most common use case could be a network issue, for which we want retries.